python/ray/experimental/raysort/main.py

```diff
@@ -190,146 +208,270 @@ def _dummy_sort_and_partition(part: np.ndarray,
+
+    pinfo = _part_info(args, reducer_id, "output")
+    _merge_impl(args, M, pinfo, get_block, args.skip_output)
+    return pinfo
+
+
+def _node_res(node: str) -> Dict[str, float]:
+    return {"resources": {f"node:{node}": 1e-3}}
+
+
+def _get_placement_groups(args: Args) -> List[ray.PlacementGroupID]:
+    """
+    TODO: document
+    """
+    worker_share = args.num_workers / args.num_reducers
+    pgs = [
+        ray.util.placement_group([{
+            "worker": worker_share
+        }]) for _ in range(args.num_reducers)
+    ]
+    ray.get([pg.ready() for pg in pgs])
+    return pgs
+
+
+@tracing_utils.timeit("sort", report_time=True)
+def sort_main(args: Args):
+    parts = _load_manifest(args, constants.INPUT_MANIFEST_FILE)
+    assert len(parts) == args.num_mappers
     boundaries = sortlib.get_boundaries(args.num_reducers)
-    mapper_results = np.empty(
-        (args.num_mappers, args.num_reducers), dtype=object)
-    for part_id, node, path in partitions:
-        opt = {} if args.skip_input else {
-            "resources": {
-                f"node:{node}": 1 / args.num_mappers
-            },
-            "memory": args.input_part_size * 1.2,
-        }
-        opt.update(num_returns=args.num_reducers)
-        mapper_results[part_id, :] = mapper.options(**opt).remote(
-            boundaries, part_id, path)
-
-    reducer_results = []
-    for r in range(args.num_reducers):
-        opt = {
-            "memory": args.output_part_size * 1.0,
-        }
-        blocks = mapper_results[:, r].tolist()
-        ret = reducer.options(**opt).remote(r, *blocks)
-        reducer_results.append(ret)
+    pgs = _get_placement_groups(args)
+
+    mapper_opt = {"num_returns": args.num_reducers}
+    merge_results = np.empty(
+        (args.num_rounds, args.num_reducers), dtype=object)
+
+    part_id = 0
+    for round in range(args.num_rounds):
+        # Limit the number of in-flight rounds.
+        num_extra_rounds = round - args.num_concurrent_rounds + 1
+        if num_extra_rounds > 0:
+            ray.wait(
+                [f for f in merge_results.flatten() if f is not None],
+                num_returns=num_extra_rounds * args.num_reducers)
+
+        # Submit map tasks.
+        mapper_results = np.empty(
+            (args.num_mappers_per_round, args.num_reducers), dtype=object)
+        for _ in range(args.num_mappers_per_round):
+            _, node, path = parts[part_id]
+            m = part_id % args.num_mappers_per_round
+            if not args.skip_input:
+                mapper_opt.update(_node_res(node))
+            mapper_results[m, :] = mapper.options(**mapper_opt).remote(
+                args, boundaries, path)
+            part_id += 1
+
+        # Submit merge tasks.
+        merge_results[round, :] = [
+            merge_mapper_blocks.options(placement_group=pgs[r]).remote(
+                args, r, round, *mapper_results[:, r].tolist())
+            for r in range(args.num_reducers)
+        ]
```

Comment:
Probably a good idea to clear the local references to mapper_results here, too, before we do the `ray.wait`.